bin/plot_krun_results

```diff
@@ -238,36 +268,74 @@ def add_inset_to_axis(axis, rect):
     return fig.add_axes([fig_left, fig_bottom, fig_width, fig_height])
 
 
-def draw_subplot(axis, data, cycles_data, title, x_range, y_range, window_size, outliers,
-                 unique, common, changepoints, changepoint_means, median, tukey):
-    data_narray = numpy.array(data)
-    cycles_data_narray = numpy.array(cycles_data)
-
-    # Plot the original measurements.
-    axis.plot(data_narray, label='Measurement', color=LINE_COLOUR,
-              zorder=ZORDER_DATA)
-    # Mark outliers, optionally.
-    if outliers:
-        axis.scatter(outliers, data_narray[outliers], color=OUTLIER_COLOR,
-                     marker=OUTLIER_MARKER, s=OUTLIER_SIZE, label='Outliers',
+def draw_process_exec(grid_cell, data, cycles_data, aperf_data, mperf_data,
+                 instr_data, title, x_bounds, y_range, window_size,
+                 outliers, unique, common, changepoints, changepoint_means,
+                 median, tukey):
+    iterations = numpy.array(xrange(x_bounds[0], x_bounds[1]))
+    # Marshal data into numpy arrays.
+    data_narray = numpy.array(data[x_bounds[0]:x_bounds[1]])
+    cycle_data_narrays = list()
+    if cycles_data:
+        cycles_min, cycles_max = float('inf'), float('-inf')
+        for core in xrange(len(cycles_data)):
+            cycles_min = min(min(cycles_data[core][x_bounds[0]:x_bounds[1]]), cycles_min)
+            cycles_max = max(max(cycles_data[core][x_bounds[0]:x_bounds[1]]), cycles_max)
+        for core in xrange(len(cycles_data)):  # For each core.
+            cycle_data_narrays.append(numpy.array(cycles_data[core]))
+    perf_data_narrays = list()
+    if aperf_data and mperf_data:
+        for core in xrange(len(aperf_data)):  # For each core.
+            perf_data_narrays.append(numpy.array([float(aperf_data[core][i]) / float(mperf_data[core][i])
+                                     for i in xrange(x_bounds[0], x_bounds[1])]))
+    instr_data_narrays = list()
+    if instr_data:
+        for chart_data in instr_data:
+            instr_data_narrays.append(numpy.array(chart_data.data))
+    # Calculate the number of plots needed.
+    n_rows = 1
+    if cycles_data:
+        n_rows += len(cycle_data_narrays)
+    if aperf_data:
+        n_rows += len(perf_data_narrays)
+    if instr_data:
+        n_rows += len(instr_data_narrays)
+    inner_grid = gridspec.GridSpecFromSubplotSpec(n_rows, 1, grid_cell, hspace=0.2)
+    # Collect handles and labels, so that we can draw one legend per page.
+    handles_labels = [list(), list()]
+    # Plot wall clock times.
+    axis1 = pyplot.subplot(inner_grid[n_rows - 1, 0])
+    axis1.plot(iterations, data_narray, label='Measurement', color=LINE_COLOUR,
+               zorder=ZORDER_DATA)
+    # If there are outliers, highlight them.
```

Comment:
There seems to be a common function here which is something like:

```python
def f(l, x_bounds):
    out1 = [val for val in l if val >= x_bounds[0] and val < x_bounds[1]]
￼    out2 = [val - x_bounds[0] for val in out1 if val >= x_bounds[0] and val < x_bounds[1]]
    return (out1, out2)
```

If I've got that right, then it's probably best separated out.